### PR TITLE
Linear phosphorylation

### PR DESCRIPTION
**Pull request by NicholasCowie**
# Summary
Phosphorylation is a key metabolic regulator where kinases and phosphatases regulate the amount of enzyme available for catalysis. Due to the sometimes ambiguous roles that kinases and phosphatases play (e.g. occasionally phosphatases can inhibit rather than activate); these terms are avoided in the code and are referred to as activators and inhibitors instead.

Future updates will allow for allosteric regulation of phosphorylation rates, however this will be included as a separate PR.
There is an example of phosphorylation included in the file `tests/data/linear_phos.toml`, which includes activation of the second reaction (This should not impact the final sampling as there is no inhibition)

# Commits

* Docs: updating name clash and including phosphorylation reasoning
* doc: removing previous name
* doc: fixing colours on plot
* feat: including phosphorylation in data structure
* feat: including phosphorylation in stan code
* style: removing unused code
* feat: retrieving information and updating inputs to stan model
* feat: parsing inputs from toml to mi object
* style: satisfying make qa
* tests: updating tests with new field
* bugs: fixing naming and logic issues with phos parsing
* data: example phosphorylation file
* tests: satisfying black styling

### CODE REVIEW

File: docs/adr/0014-phosphorylation-as-mass-action.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+There are many potential regulatory mechanisms to implement, as discussed in
+https://doi.org/10.1016/j.biosystems.2005.05.015. with the following assumptions
+the selected rate law is the linear rate law:
+
+* [Kinase] and [Phosphotase] << [Metabolic Enzyme],
+* Rapid equilibrium binding between phosphorylation enzymes and metabolic enzyme,
+* [Kinase] << dissociation constant for Kinase,
+* [Phosphotase] << dissociation constant for Phosphotse,
+* Competitive binding of phosphotases and kinases is negligable.
+* The ATP/ADP ratio
```
Comment (teddygroves) [inline docs/adr/0014-phosphorylation-as-mass-action.md:28]: Incomplete sentence

File: src/maud/data_model.py [hunk 2]
```diff
@@ -190,6 +190,29 @@ def __init__(
         self.stoichiometry = stoichiometry
 
 
+class Phosphorylation:
+    """Constructor for the reaction object.
```
Comment (teddygroves) [inline src/maud/data_model.py:194]: Update description

File: src/maud/data_model.py [hunk 3]
```diff
@@ -190,6 +190,29 @@ def __init__(
...
+
+    :param id: drain id, use a BiGG id if possible.
+    :param name: drain name.
+    :param stoichiometry: reaction stoichiometry,
```
Comment (teddygroves) [inline src/maud/data_model.py:198]: These are out of date I think

File: src/maud/data_model.py [hunk 4]
```diff
@@ -190,6 +190,29 @@ def __init__(
...
+    """
+
+    def __init__(
+        self,
+        id: str,
+        name: str = None,
+        activating: bool = None,
+        inhibiting: bool = None,
+        enzyme_id: str = None,
```
Comment (teddygroves) [inline src/maud/data_model.py:207]: Are you sure these things should be allowed to be `None`?

File: src/maud/dbalanced_dt.stan [hunk 5]
```diff
@@ -21,6 +21,15 @@ int[] get_enz_mics(matrix S, int i_enz){
   return out;
 }
 
+real get_active_enzyme_fraction(vector activating_enzyme_conc,
+                                vector deactivating_enzyme_conc,
+                                vector phosphorylating_enzyme_kcat){
+  real alpha = sum(phosphorylating_enzyme_kcat .* phosphorylating_enzyme_kcat);
+  real beta = sum(phosphorylating_enzyme_kcat .* phosphorylating_enzyme_kcat);
```
Comment (teddygroves) [inline src/maud/dbalanced_dt.stan]: These are the same - is that right?

File: src/maud/inference_model.stan [hunk 6]
```diff
@@ -15,6 +15,7 @@ data {
   int<lower=1> N_reaction;
   int<lower=1> N_enzyme;
   int<lower=0> N_drain;
+  int<lower=0> N_phophorylation_enzymes;
```
Comment (teddygroves) [inline src/maud/inference_model.stan:18]: typo (and below): phophorylation

File: src/maud/io.py [hunk 7]
```diff
@@ -45,26 +46,6 @@
 from maud.utils import codify
 
 
-MECHANISM_TO_PARAM_IDS = {
```
Comment (teddygroves) [inline src/maud/io.py:48]: Don't know how this survived so long!

File: src/maud/io.py [hunk 8]
```diff
@@ -135,12 +128,13 @@ def get_stan_codes(km: KineticModel, experiments: ExperimentSet) -> StanCodeSet:
...
         experiment_codes=codify([e.id for e in experiments.experiments]),
         enzyme_codes=codify(enzyme_ids),
+        phos_enz_codes=phos_enz_codes,
         drain_codes=drain_codes,
     )
 
 
 def load_drain_from_toml(toml_drain: dict) -> Drain:
-    """Turn a dictionary representing a drain into a Drain object.
+    """Return a dictionary representing a drain into a Drain object.
```
Comment (teddygroves) [inline src/maud/io.py:137]: I think "Turn" was correct here

File: src/maud/io.py [hunk 9]
```diff
@@ -154,6 +148,28 @@ def load_drain_from_toml(toml_drain: dict) -> Drain:
     )
 
 
+def load_phosphorylation_from_toml(toml_phosphorylation: dict) -> Phosphorylation:
+    """Return a dictionary representing a drain into a drain object.
```
Comment (teddygroves) [inline src/maud/io.py:152]: Out of date I think

File: tests/data/linear_phos.toml [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+[[reactions.enzymes]]
+id = 'r3'
+name = 'the enzyme that catalyses reaction r3'
+mechanism = 'modular_rate_law'
+
+[[phosphorylation]]
+id = "p2"
+name = "phosphorylation on reaction number 2"
+enzyme_id = "r2"
+activating = 1
```
Comment (teddygroves) [inline tests/data/linear_phos.toml:76]: Better to use `True` or similar than 1

**Review by teddygroves — COMMENTED** [review]
I think I found a problem in the Stan code, otherwise mostly typos

File: src/maud/io.py (hunk 7 above)
Comment (NicholasCowie) [reply src/maud/io.py:48]: Hope that we would bring them back I suppose

**Review by teddygroves — APPROVED** [review]
Great work - looks good!

**Pull request merged**